app/models/concerns/training_period_sources.rb

```diff
@@ -0,0 +1,29 @@
+module TrainingPeriodSources
+  extend ActiveSupport::Concern
+
+  def contract_period
+    @contract_period ||= ContractPeriod.containing_date(started_on)
+  end
+
+  def active_lead_provider
+    @active_lead_provider ||= ActiveLeadProvider.find_by(
+      lead_provider:,
+      contract_period:
+    ) || raise("Missing ActiveLeadProvider for #{lead_provider&.name} in #{contract_period&.year}")
```

Comment:
Is the custom error here worth it, vs just a `find_by!`?

```suggestion
    @active_lead_provider ||= ActiveLeadProvider.find_by!(lead_provider:, contract_period:)
```